Draft: Add support for msi #24

Hugo-C opened the pull request:
Hello, I am trying to fill the gaps in https://github.com/ralphje/signify/issues/24 to properly verify an MSI's authenticode.
The parsing is done by the `olefile` lib and the various hashes are computed using the work done in https://github.com/sassoftware/relic/blob/master/lib/authenticode/msiverify.go (I could not find a specification document for authenticode on MSI/OLE file).

I thought it would be better to share what I came up with, as it is functional but not nicely integrated with the rest of the lib. The questions I have to continue are:
* Do we want a class above `SignedMsiFile`/`SignedPEFile` to reduce duplicated code ?
* On [Relic](https://github.com/sassoftware/relic) side they verify the `MsiDigitalSignatureEx` if present, is it something we also want to do ?
* I could not find a way to use `Fingerprinter`, as `olefile` is doing the parsing, is it ok or do you see a solution ?
* Do we want to add `olefile` as a dependency or keep it as an [extra](https://packaging.python.org/en/latest/tutorials/installing-packages/#installing-extras) ? 

I also need to write a proper doc inside `SignedMsiFile` and find a signed MSI with nested files.
Let me know your thoughts and others element to be done before the MR can be merged

Hugo-C committed:
Add support for msi #24

ralphje commented:
I think this is pretty neat! We should certainly integrate your work. There are some things I'd like to note first:

* You wouldn't need to add the license comment to any new files you add. This is a requirement due to splitting off from verify-sigs, but new files won't need that.
* I'd like to prevent duplicate code, so the common code would probably best be placed in some common class. Please make sure to do this in two separate commits, so that it's easier to review your changes, as I wouldn't need to review moved code :). That also makes it a bit hard to review in detail now.
* I think it's OK to use olefile for the parsing and hashing. Please make sure to include olefile as package requirement. I've just updated the repository to use pyproject.toml. Adding it as an extra makes sense to me. (We would need to make sure that we do not import code into the main project if olefile is an extra.)
* Please create a separate test class or a separate test file, to split it off a bit. I've fixed tests in GitHub, so it should properly test now if you rebase.

I will do a more in-depth review of the relevant code once the PR is cleaned up a bit.

Hugo-C committed:
Remove licence header and fix black

Hugo-C committed:
Fix ruff

Hugo-C committed:
Fix mypy

Hugo-C committed:
Add olefile as extra dependency

Hugo-C committed:
Extract msi tests in a new file

Hugo-C committed:
Raise AuthenticodeNotSignedError on invalid olefile

Hugo-C committed:
Raise AuthenticodeNotSignedError on unsigned msi

Hugo-C committed:
Add msi without extended digital signature

Hugo-C committed:
Handle directory entries in prehash

Hugo-C committed:
Handle directory entries in msi hash

Hugo-C committed:
Refactor _prehash_storage_entry to classmethod

Hugo-C committed:
Fix mypy errors

Hugo-C committed:
Fix black

Hugo-C committed:
Clean up signed_msi.py

Hugo-C committed:
Introduce SignedFile super class

Hugo-C committed:
Fix ruff

Hugo-C commented:
@ralphje can you take a look ?
I did my best to clean up the PR. I tested it on a complex MSI with [libreoffice](https://libreoffice.org/download) which I did not include as it is quite large.
I fixed mypy/black/ruff errors and rebased the branch.

I added a `SignedFile` as an interface for all signed files and to reduce a bit the duplicated code.

ralphje commented:
Awesome! To manage expectations: I'm currently out of the country, so it will take a few weeks for me to get to reviewing your contribution.

Hugo-C committed:
Fix mypy errors for 3.10 syntax

Hugo-C commented:
I have run CI on my side, so it should be OK here too.
There is no rush of course :slightly_smiling_face:

Squiblydoo committed:
Update __init__.py

(cherry picked from commit 1bf237bef3ab31583794dbf8e3c777c15fe832a6)

Squiblydoo commented:
Hi! Since it has been ~3 weeks since the last message, it seemed good to follow up. 

I'm interested in seeing the MSI support being merged as well. 

I also have some suggested improvements I've made in my own fork that I may recommend to be included if/when the MSI fork is merged.

1. For my own purposes, I also created an ["Example" for the MSI extraction](https://github.com/Squiblydoo/signify/blob/master/examples/MSI_info.py) similar to the original "authenticode_info.py" to demonstrate `signify`'s capability. 
2. For the analysis pipeline, I also created a "[RawCertificateFile](https://github.com/Squiblydoo/signify/blob/master/signify/authenticode/raw_cert.py)" class; this allows me to parse the Authenticode certificate from an EXE or MSI without loading the file a second time after it has already been loaded for analysis by other tools. (I also have an [example script](https://github.com/Squiblydoo/signify/blob/master/examples/raw_cert_parse.py) for this.)
3. I am also [augmenting the OID to RDN table](https://github.com/Squiblydoo/signify/blob/e07d991435ca67250567bd7127766c4dbb2d9011/signify/x509/certificates.py#L250). (I am planning to do more analysis shortly to identify other common OID that should be extracted, but in my code, I provide some references for OID to RDN tables.)

Hugo-C committed:
Use shorter import in tests

Hugo-C committed:
Fix ruff

Squiblydoo commented:
Ah yep, understood. Yeah, I plan to suggest the other changes in separate PR, but just wanted to say that depend a fair bit on this one. :)
